src/spikeinterface/postprocessing/template_similarity.py

```diff
@@ -60,25 +85,102 @@ def _get_data(self):
 compute_template_similarity = ComputeTemplateSimilarity.function_factory()
 
 
-def compute_similarity_with_templates_array(templates_array, other_templates_array, method):
+def compute_similarity_with_templates_array(templates_array, other_templates_array, method, n_shifts, mask=None):
+
     import sklearn.metrics.pairwise
 
-    if method == "cosine_similarity":
+    all_metrics = ["cosine", "l1", "l2"]
+
+    if method in all_metrics:
+        nb_templates = templates_array.shape[0]
         assert templates_array.shape[0] == other_templates_array.shape[0]
-        templates_flat = templates_array.reshape(templates_array.shape[0], -1)
-        other_templates_flat = templates_array.reshape(other_templates_array.shape[0], -1)
-        similarity = sklearn.metrics.pairwise.cosine_similarity(templates_flat, other_templates_flat)
+        n = templates_array.shape[1]
+        nb_templates = templates_array.shape[0]
+        assert n_shifts < n, "max_lag is too large"
+        num_shifts = 2 * n_shifts + 1
+        distances = np.ones((num_shifts, nb_templates, nb_templates), dtype=np.float32)
+        if mask is not None:
+            units_overlaps = np.sum(mask, axis=2) > 0
+            overlapping_templates = {}
+            for i in range(nb_templates):
+                overlapping_templates[i] = np.flatnonzero(units_overlaps[i])
+
+        # We can use the fact that dist[i,j] at lag t is equal to dist[j,i] at time -t
+        # So the matrix can be computed only for negative lags and be transposed
+
+        for count, shift in enumerate(range(-n_shifts, 1)):
+            if mask is None:
```

Comment:
We should not duplicate the code twice when mask is None but create a mask that contain a full intersection. No ?

Reply:
done